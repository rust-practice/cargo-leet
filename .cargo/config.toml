[alias]
i = "install --path . --features=tool"
lg = "run --features=tool -- leet gen"
