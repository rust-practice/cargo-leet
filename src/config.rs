pub struct Config {}

impl Config {
    pub const LEETCODE_URL: &str = "https://leetcode.com/problems/";
    pub const LEETCODE_GRAPH_QL: &str = "https://leetcode.com/graphql/";
}
