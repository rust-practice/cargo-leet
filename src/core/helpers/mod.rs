pub(crate) mod code_snippet;
pub(crate) mod daily_challenge;
pub(crate) mod problem_metadata;
pub(crate) mod write_to_disk;
