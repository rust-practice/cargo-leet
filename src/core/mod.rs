mod generate;
mod helpers;

use self::generate::do_generate;
use crate::cli::Cli;
use anyhow::{bail, Context};
use std::{env, path::Path};

pub fn run(cli: &Cli) -> anyhow::Result<()> {
    cli.update_current_working_dir()?;

    working_directory_validation()?;

    match &cli.command {
        crate::cli::Commands::Generate(args) => do_generate(args),
    }
}

fn working_directory_validation() -> anyhow::Result<()> {
    let req_file = "Cargo.toml";
    let path = Path::new(req_file);
    if path.exists() {
        Ok(())
    } else {
        bail!(
            "Failed to find {req_file} in current directory '{}'",
            env::current_dir()
                .context("Failed to get current working directory")?
                .display()
        );
    }
}
