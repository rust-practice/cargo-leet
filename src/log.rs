use anyhow::Context;
use log::LevelFilter;
use log4rs::{
    append::console::ConsoleAppender,
    config::{Appender, Root},
    encode::pattern::PatternEncoder,
};

pub fn init_logging(log_level: LevelFilter) -> anyhow::Result<()> {
    let stdout = ConsoleAppender::builder()
        .encoder(Box::new(PatternEncoder::new("{h({M} {l} - {m})}{n}")))
        .build();

    let config = log4rs::Config::builder()
        .appender(Appender::builder().build("stdout", Box::new(stdout)))
        .build(Root::builder().appender("stdout").build(log_level))?;
    log4rs::init_config(config).context("Failed to initialize logging")?;
    Ok(())
}
