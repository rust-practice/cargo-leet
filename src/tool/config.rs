pub(crate) struct Config {}

impl Config {
    // assumed in the code using them URLs Must end with trailing "/"
    pub(crate) const LEETCODE_PROBLEM_URL: &str = "https://leetcode.com/problems/";
    pub(crate) const LEETCODE_GRAPH_QL: &str = "https://leetcode.com/graphql/";
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn problem_url_ends_with_slash() {
        assert!(Config::LEETCODE_PROBLEM_URL.ends_with('/'));
    }

    #[test]
    fn graph_ql_url_ends_with_slash() {
        assert!(Config::LEETCODE_GRAPH_QL.ends_with('/'));
    }
}
