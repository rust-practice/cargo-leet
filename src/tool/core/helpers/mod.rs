pub(crate) mod code_snippet;
pub(crate) mod daily_challenge;
pub(crate) mod local_store;
pub(crate) mod problem_code;
pub(crate) mod problem_metadata;
pub(crate) mod write_to_disk;
