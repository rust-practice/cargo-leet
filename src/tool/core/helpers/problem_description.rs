use crate::tool::config::Config;
use anyhow::Context;
use log::info;

use super::{get_response, local_store::path_local_store_problem_description};

pub(crate) mod data_structure;

pub(crate) fn get_problem_description(
    title_slug: &str,
) -> anyhow::Result<data_structure::ProblemDescription> {
    info!("Attempting to get problem description");
    get_problem_description_response(title_slug)?.try_into()
}

fn get_problem_description_response(
    title_slug: &str,
) -> anyhow::Result<data_structure::ProblemDescriptionResponse> {
    get_response(
        title_slug,
        local_store_request_problem_description,
        external_request_problem_description,
    )
}

fn local_store_request_problem_description(title_slug: &str) -> anyhow::Result<String> {
    let path = path_local_store_problem_description(title_slug);
    std::fs::read_to_string(&path).with_context(|| format!("failed to read string from {path:?}"))
}

fn external_request_problem_description(title_slug: &str) -> anyhow::Result<String> {
    info!("[External] Going to send request for problem description for problem with title: {title_slug}");
    ureq::get(Config::LEETCODE_GRAPH_QL)
        .send_json(ureq::json!({
            "query": r"query questionContent($titleSlug: String!) {
            question(titleSlug: $titleSlug) {
                content
            }
        }",
            "variables":{"titleSlug": title_slug},
            "operationName":"questionContent"
        }))
        .context("failed to get request for description failed")?
        .into_string()
        .context("failed to convert response into String")
}

#[cfg(test)]
mod tests {
    use std::io::Write as _;

    use anyhow::Context;
    use rstest::rstest;

    use crate::tool::core::helpers::{
        local_store::{
            path_local_store_problem_description,
            tests::{get_rnd_request_delay, insta_settings, title_slugs, SlugList},
        },
        problem_description::{external_request_problem_description, get_problem_description},
    };

    #[rstest]
    #[ignore = "Only use for downloading responses"]
    fn download_response_from_leetcode(title_slugs: SlugList) {
        for title_slug in title_slugs {
            let sleep_delay = std::time::Duration::from_millis(get_rnd_request_delay());
            println!(
            "Going to sleep for {} milliseconds before requesting and trying to save {title_slug}",
            sleep_delay.as_millis()
        );
            std::thread::sleep(sleep_delay); // Sleep to not go too hard on leetcode API
            let response_string = external_request_problem_description(title_slug).unwrap();
            let path = path_local_store_problem_description(title_slug);
            let mut file = std::fs::OpenOptions::new()
                .write(true)
                .create(true)
                .truncate(true)
                .open(&path)
                .with_context(|| format!("failed to save response to {path:?}"))
                .unwrap();
            file.write_all(response_string.as_bytes()).unwrap();
            println!("Save of {title_slug:?} completed\n");
        }
        println!("Successfully saved all responses");
    }

    #[rstest]
    fn conversion_from_leetcode_response(title_slugs: SlugList, insta_settings: insta::Settings) {
        for title_slug in title_slugs {
            insta_settings.bind(|| {
                insta::assert_debug_snapshot!(get_problem_description(title_slug).unwrap());
            });
        }
    }

    #[rstest]
    fn extract_solutions_from_description(title_slugs: SlugList, insta_settings: insta::Settings) {
        for title_slug in title_slugs {
            insta_settings.bind(|| {
                let problem_description = get_problem_description(title_slug).unwrap();
                insta::assert_debug_snapshot!(problem_description.get_solutions());
            });
        }
    }
}
