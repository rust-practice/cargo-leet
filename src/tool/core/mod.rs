mod generate;
mod helpers;

use self::generate::do_generate;
use crate::tool::cli::{self, Cli};
use anyhow::{bail, Context};
use std::{env, path::Path};

/// Entry point used by the tool. The `main.rs` is pretty thin shim around this
/// function.
///
/// # Errors
/// - Current directory does not exist.
/// - There are insufficient permissions to access the current directory.
/// - Returns an Err if the operation fails.
pub fn run(cli: &Cli) -> anyhow::Result<()> {
    cli.update_current_working_dir()?;

    working_directory_validation()?;

    match &cli.command {
        cli::Commands::Generate(args) => do_generate(args),
    }
}

fn working_directory_validation() -> anyhow::Result<()> {
    let req_file = "Cargo.toml";
    let path = Path::new(req_file);
    if path.exists() {
        Ok(())
    } else {
        bail!(
            "Failed to find {req_file} in current directory '{}'",
            env::current_dir()
                .context("Failed to get current working directory")?
                .display()
        );
    }
}
