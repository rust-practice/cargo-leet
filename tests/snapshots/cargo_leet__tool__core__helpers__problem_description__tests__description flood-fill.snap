---
source: src/tool/core/helpers/problem_description.rs
expression: get_problem_description(title_slug).unwrap()
---
ProblemDescription {
    content: "<p>You are given an image represented by an <code>m x n</code> grid of integers <code>image</code>, where <code>image[i][j]</code> represents the pixel value of the image. You are also given three integers <code>sr</code>, <code>sc</code>, and <code>color</code>. Your task is to perform a <strong>flood fill</strong> on the image starting from the pixel <code>image[sr][sc]</code>.</p>\n\n<p>To perform a <strong>flood fill</strong>:</p>\n\n<ol>\n\t<li>Begin with the starting pixel and change its color to <code>color</code>.</li>\n\t<li>Perform the same process for each pixel that is <strong>directly adjacent</strong> (pixels that share a side with the original pixel, either horizontally or vertically) and shares the <strong>same color</strong> as the starting pixel.</li>\n\t<li>Keep <strong>repeating</strong> this process by checking neighboring pixels of the <em>updated</em> pixels&nbsp;and modifying their color if it matches the original color of the starting pixel.</li>\n\t<li>The process <strong>stops</strong> when there are <strong>no more</strong> adjacent pixels of the original color to update.</li>\n</ol>\n\n<p>Return the <strong>modified</strong> image after performing the flood fill.</p>\n\n<p>&nbsp;</p>\n<p><strong class=\"example\">Example 1:</strong></p>\n\n<div class=\"example-block\">\n<p><strong>Input:</strong> <span class=\"example-io\">image = [[1,1,1],[1,1,0],[1,0,1]], sr = 1, sc = 1, color = 2</span></p>\n\n<p><strong>Output:</strong> <span class=\"example-io\">[[2,2,2],[2,2,0],[2,0,1]]</span></p>\n\n<p><strong>Explanation:</strong></p>\n\n<p><img alt=\"\" src=\"https://assets.leetcode.com/uploads/2021/06/01/flood1-grid.jpg\" style=\"width: 613px; height: 253px;\" /></p>\n\n<p>From the center of the image with position <code>(sr, sc) = (1, 1)</code> (i.e., the red pixel), all pixels connected by a path of the same color as the starting pixel (i.e., the blue pixels) are colored with the new color.</p>\n\n<p>Note the bottom corner is <strong>not</strong> colored 2, because it is not horizontally or vertically connected to the starting pixel.</p>\n</div>\n\n<p><strong class=\"example\">Example 2:</strong></p>\n\n<div class=\"example-block\">\n<p><strong>Input:</strong> <span class=\"example-io\">image = [[0,0,0],[0,0,0]], sr = 0, sc = 0, color = 0</span></p>\n\n<p><strong>Output:</strong> <span class=\"example-io\">[[0,0,0],[0,0,0]]</span></p>\n\n<p><strong>Explanation:</strong></p>\n\n<p>The starting pixel is already colored with 0, which is the same as the target color. Therefore, no changes are made to the image.</p>\n</div>\n\n<p>&nbsp;</p>\n<p><strong>Constraints:</strong></p>\n\n<ul>\n\t<li><code>m == image.length</code></li>\n\t<li><code>n == image[i].length</code></li>\n\t<li><code>1 &lt;= m, n &lt;= 50</code></li>\n\t<li><code>0 &lt;= image[i][j], color &lt; 2<sup>16</sup></code></li>\n\t<li><code>0 &lt;= sr &lt; m</code></li>\n\t<li><code>0 &lt;= sc &lt; n</code></li>\n</ul>\n",
}
